lib/hex/hex.cpp

```diff
@@ -0,0 +1,119 @@
+// EVMC: Ethereum Client-VM Connector API.
+// Copyright 2021 The EVMC Authors.
+// Licensed under the Apache License, Version 2.0.
+
+#include <evmc/hex.hpp>
+#include <cctype>
+
+namespace evmc
+{
+namespace
+{
+struct hex_error_category : std::error_category
+{
+    const char* name() const noexcept override { return "hex"; }
+
+    std::string message(int ev) const override
+    {
+        switch (static_cast<hex_errc>(ev))
+        {
+        case hex_errc::invalid_hex_digit:
+            return "invalid hex digit";
+        case hex_errc::incomplete_hex_byte_pair:
+            return "incomplete hex byte pair";
+        default:
+            return "unknown error";
+        }
+    }
+};
+
+inline int from_hex_digit(char h)
+{
+    if (h >= '0' && h <= '9')
+        return h - '0';
+    else if (h >= 'a' && h <= 'f')
+        return h - 'a' + 10;
+    else if (h >= 'A' && h <= 'F')
+        return h - 'A' + 10;
+    else
+        throw hex_error{hex_errc::invalid_hex_digit};
+}
+
+template <typename OutputIt>
+inline void from_hex(const std::string& hex, OutputIt result)
+{
+    // TODO: This can be implemented with hex_decode_iterator and std::copy.
+
+    // Omit the optional 0x prefix.
+    const auto hex_begin =
+        (hex.size() >= 2 && hex[0] == '0' && hex[1] == 'x') ? hex.begin() + 2 : hex.begin();
+
+    constexpr int empty_byte_mark = -1;
+    int b = empty_byte_mark;
+    for (auto it = hex_begin; it != hex.end(); ++it)
+    {
+        const auto h = *it;
+        if (std::isspace(h))
+            continue;
+
+        const int v = from_hex_digit(h);
+        if (b == empty_byte_mark)
+        {
+            b = v << 4;
+        }
+        else
+        {
+            *result++ = static_cast<uint8_t>(b | v);
+            b = empty_byte_mark;
+        }
+    }
+
+    if (b != empty_byte_mark)
+        throw hex_error{hex_errc::incomplete_hex_byte_pair};
+}
+}  // namespace
+
+std::error_code make_error_code(hex_errc errc) noexcept
+{
+    static hex_error_category category;
```

Comment:
```suggestion
    static const hex_error_category category;
```

Reply:
Move to `hex_category()` with `static const`.